File: Construction-SIEM-Platform/backend/app/routers/threat_intel.py
from fastapi import APIRouter, Query

from app.schemas.threat_intel import (
    CorrelateEventRequest,
    CorrelateEventResponse,
    IoCCheckRequest,
    IoCCheckResponse,
    RecentThreatsResponse,
)
from app.services.threat_intel import get_threat_intel_service

router = APIRouter(prefix="/api/v1/threat-intel", tags=["threat-intel"])


@router.post("/check", response_model=IoCCheckResponse)
async def check_ioc(payload: IoCCheckRequest):
    service = get_threat_intel_service()
    result = service.check_ioc(value=payload.value, ioc_type=payload.ioc_type)
    return IoCCheckResponse(**result)


@router.post("/correlate", response_model=CorrelateEventResponse)
async def correlate_event(payload: CorrelateEventRequest):
    service = get_threat_intel_service()
    matches = service.correlate_event_with_iocs(payload.event_data)
    risk_score = 0.0
    if matches:
        risk_score = sum(
            m["match"]["confidence"] * 7.0
            for m in matches
        ) / len(matches)
    return CorrelateEventResponse(
        status="success",
        matches=matches,
        total_matches=len(matches),
        risk_score=round(risk_score, 2),
    )


@router.get("/recent", response_model=RecentThreatsResponse)
async def get_recent_threats(hours: int = Query(24, ge=1, le=720)):
    service = get_threat_intel_service()
    recent = service.get_recent_threats(hours=hours)
    return RecentThreatsResponse(data=recent)

File: Construction-SIEM-Platform/frontend/src/api/threatIntel.ts
import client from './client'
import type {
  IoCCheckRequest,
  IoCCheckResponse,
  CorrelateEventRequest,
  CorrelateEventResponse,
  RecentThreatsResponse,
} from '@/types'

export const threatIntelApi = {
  check(payload: IoCCheckRequest): Promise<IoCCheckResponse> {
    return client.post<IoCCheckResponse>('/threat-intel/check', payload).then((r) => r.data)
  },

  correlate(payload: CorrelateEventRequest): Promise<CorrelateEventResponse> {
    return client
      .post<CorrelateEventResponse>('/threat-intel/correlate', payload)
      .then((r) => r.data)
  },

  recent(hours = 24): Promise<RecentThreatsResponse> {
    return client
      .get<RecentThreatsResponse>('/threat-intel/recent', { params: { hours } })
      .then((r) => r.data)
  },
}
